### PR TITLE
Fixing the statistics for MultiRunHarvesting

### PR DESCRIPTION
**Pull request by dimattia**
Greetings,

the main purpose of this pull request is to raise the threshold on the total amount of clusters that have to be collected before fitting the APV cluster charge thus producing the gain payload. The change consists of a setting of a parameter in two python filles. 

Since the cluster statistics affects the precision of the gain payload, it is important to have this PR merged in the main branch and to have it delivered at Tier0 asap. indeed it impacts the gain computation in  the MRH framework. The framework that uses the calibration ntuple is not affected by this one, because - in this case - the input statistic is handled manually.

In addition to the statistic fix, the fill of some monitoring histogram has been protected against the inclusion of statistics from the Pixel detector. This change concerns the inclusion of an if statement into two .cc files.

### CODE REVIEW

**Comment by arunhep** [comment]
please test

**Comment by mmusich** [comment]
The errors that show up in most tested workflows:
```
Traceback (most recent call last):
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/bin/slc6_amd64_gcc530/cmsDriver.py", line 55, in <module>
    run()
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/bin/slc6_amd64_gcc530/cmsDriver.py", line 27, in run
    configBuilder.prepare()
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/python/Configuration/Applications/ConfigBuilder.py", line 2052, in prepare
    self.addStandardSequences()
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/python/Configuration/Applications/ConfigBuilder.py", line 744, in addStandardSequences
    getattr(self,"prepare_"+stepName)(sequence = getattr(self,stepName+"DefaultSeq"))
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/python/Configuration/Applications/ConfigBuilder.py", line 1642, in prepare_RECO
    self.loadDefaultOrSpecifiedCFF(sequence,self.RECODefaultCFF)
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/python/Configuration/Applications/ConfigBuilder.py", line 1185, in loadDefaultOrSpecifiedCFF
    l=self.loadAndRemember(defaultCFF)
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/python/Configuration/Applications/ConfigBuilder.py", line 321, in loadAndRemember
    self.process.load(includeFile)
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/python/FWCore/ParameterSet/Config.py", line 582, in load
    module = __import__(moduleName)
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/python/FastSimulation/Configuration/Reconstruction_AftMix_cff.py", line 14, in <module>
    import Configuration.StandardSequences.Reconstruction_cff as _reco
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/python/Configuration/StandardSequences/Reconstruction_cff.py", line 35, in <module>
    from RecoLocalTracker.Configuration.RecoLocalTracker_cff import *
  File "/cvmfs/cms-ib.cern.ch/nweek-02477/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_2_X_2017-06-21-1100/python/RecoLocalTracker/Configuration/RecoLocalTracker_cff.py", line 21, in <module>
    from RecoLocalTracker.Phase2ITPixelClusterizer.Phase2ITPixelClusterizer_cfi import *
ImportError: No module named Phase2ITPixelClusterizer.Phase2ITPixelClusterizer_cfi
```
are not related tot the specific code changes in this PR.
In fact the IB used to perform the tests (`CMSSW_9_2_X_2017-06-21-1100`) appears to be broken: https://cms-sw.github.io/relvalLogDetail.html#slc7_aarch64_gcc530;CMSSW_9_2_X_2017-06-21-2300 
@smuzaffar

**Comment by boudoul** [comment]
It will be fixed in the next IB

**Comment by cerminar** [comment]
please test

**Comment by mmusich** [comment]
@cerminar, thanks, but if I am not mistaken reading the jenkins console output https://cmssdt.cern.ch/jenkins/job/ib-any-integration/20829/console
we're still being tested agains CMSSW_9_2_X_2017-06-21-2300, while the first IB with the fix from @boudoul is CMSSW_9_2_X_2017-06-22-1000

**Comment by smuzaffar** [comment]
please abort
10h00 is not yet ready. I will trigger the Pr tests for this once the IB is ready.

**Comment by smuzaffar** [comment]
please test

**Comment by arunhep** [comment]
+1

**Comment by cerminar** [comment]
@dimattia the increase that was discussed in the e-mail exchange was of a factor 2 while here you increased of a factor 8 the number of clusters.

I have doubts about the feasibility of this calibration in multirun mode: opening a factor 8 the current # of files might result in a job with very little chances of succeeding.

What motivates the increase? Is this just the statistics of the fit? Would a rebinning of the histograms help. Do you have any study showing the impact of the binning on the precision of the measurement?

Thanks,
G

**Comment by mmusich** [comment]
Hi @cerminar,

> Would a rebinning of the histograms help. Do you have any study showing the impact of the binning on the precision of the measurement?

in fact we do: 

https://indico.cern.ch/event/624099/contributions/2633519/attachments/1480835/2296262/170607_-_Gain2Status.pdf

(though the study has been carried out on an experimental payload with 2017 data, not yet deployed). 
There is preliminary evidence that reducing the number of bins would actually be safe and would as a matter of facts (as one might expect from first principles) reduce the average error on the landau fits.
IMHO reducing the number of bins per APV would have a double advantage of:

   * reducing the memory footprint when running the workflow multi-threaded
   * ensure that an higher number of landau fits succeeds with lower statistical fit errors even when using a reduced input statistics.

my 2 cents.

**Comment by dimattia** [comment]
@cerminar 

The increase of statistics is motivated by achieving the needed precision in the gain calibration and to have almost all Landau fits done correctly. This amount of clusters statistics is very well established through 2015 and 2016 and - as you can see from here (as example) - 

https://test-stripcalibvalidation.web.cern.ch/test-stripcalibvalidation/CalibrationValidation/ParticleGain/Run_247252_to_247988_PCL/

we always managed to calibrate the detector even at the startup of the data taking.

Concerning the histogram rebinning, as Marco said, there is a new study that can help in reducing the statistics but it still needs to be validated.

File: CalibTracker/SiStripChannelGain/plugins/SiStripGainFromCalibTree.cc [hunk 1]
```diff
@@ -1360,18 +1360,19 @@ void SiStripGainFromCalibTree::qualityMonitor() {
...
-            if(Thickness<0.04) MPVs320->Fill(Phi,FitMPV);
-            if(Thickness>0.04) MPVs500->Fill(Phi,FitMPV);
-
-            MPVError->Fill(FitMPVErr);
-            MPVErrorVsMPV->Fill(FitMPV,FitMPVErr);
-            MPVErrorVsEta->Fill(Eta,FitMPVErr);
-            MPVErrorVsPhi->Fill(Phi,FitMPVErr);
-            MPVErrorVsN->Fill(NEntries,FitMPVErr);
-
+            if ( SubDet>=3 ) {  // remove input from Pixel detector id.
```
Comment (davidlange6) [inline CalibTracker/SiStripChannelGain/plugins/SiStripGainFromCalibTree.cc]: could you not just skip SubDet <3 from the start of the loop? (eg, continue after line 1329?)

**Comment by dimattia** [comment]
@ davidlange6

I implemented your suggestion as it improves the readability of the code.

**Comment by arunhep** [comment]
please test

**Comment by arunhep** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**